handle/Timer.lua

```diff
@@ -1,37 +1,210 @@
+replaceNative('TimerGetTimeout', function(t) return Timer[GetHandleId(t)].duration end)
+replaceNative('TimerGetElapsed', function(t) return Native.TimerGetElapsed(t) + Timer[GetHandleId(t)].elapsed end)
+
+timerCallback = function()
+	local tr 	= GetExpiredTimer()
+	local ir	= GetHandleId(tr)
+
+	Timer[ir].inCallback = true
+	Timer[ir].running    = false
+	Timer[ir].callback()
+
+	Timer[ir].inCallback = false
+	if Timer[ir].onDestroyFlag then
+		DestroyTimer(tr)
+	elseif Timer[ir].tempData then
+		--  Properties of the timer were overwritten
+		Native.PauseTimer(tr)
+		TimerStart(tr, Timer[ir].tempData.dur, Timer[ir].tempData.looper, Timer[ir].tempData.func)
+		Timer[ir].tempData 	= nil
+		Timer[ir].elapsed 	= 0.
+	else
+		if Timer[ir].looped then
+			if ((BlzBitAnd(Timer[ir].pauseFlag, FLAG_TIMER_PAUSED) ~= 0) or (BlzBitAnd(Timer[ir].pauseFlag, FLAG_TIMER_ALTERED) ~= 0)) then
+				Timer[ir].pauseFlag  = 0
+				Native.PauseTimer(tr)
+				TimerStart(tr, Timer[ir].duration, Timer[ir].looped, Timer[ir].callback)
+			else
+				Timer[ir].running    = true
+			end
+		elseif BlzBitAnd(Timer[ir].pauseFlag, FLAG_TIMER_RESUMED) ~= 0 then
+			print('Timer was resumed')
+			Timer[ir].pauseFlag  = 0
+			TimerStart(tr, Timer[ir].duration, Timer[ir].looped, Timer[ir].callback)
+		end
+		Timer[ir].elapsed = 0.
+	end
+end
+
+-- @param t timer
+-- @param newR new remaining
+-- @param update update the total duration
+function TimerSetRemaining(t, newR, update)
+	local i = GetHandleId(t)
+	if Timer[i].inCallback then
+		Log.warn('TimerSetRemaining: The remaining duration of the timer ' .. I2S(i) .. ' cannot be altered while the callback function is running.')
+		return false
+	elseif not Timer[i].hasCallback then
+		Log.warn('TimerSetRemaining: The remaining duration of the timer ' .. I2S(i) .. ' cannot be altered while the timer has not yet started.')
+		return false
+	end
+	--  Ensure that newR is not below 0.       
+	newR                = math.max(newR, 0.)
+	local epsilon       = Native.TimerGetElapsed(t)
+	local delta         = TimerGetRemaining(t) - newR
+	local newDur        = Timer[i].duration - delta
+	--  If delta is anything but 0., continue with TimerSetRemaining
+	if delta ~= 0. then
+		Native.PauseTimer(t)
+		Native.TimerStart(t, newR, Timer[i].looped, Timer[i].callback)
+		Timer[i].elapsed    = Timer[i].elapsed + epsilon
+		if update then
+			Timer[i].duration   = newDur
+		else
+			Timer[i].duration	= newDur + delta
+		end
+		Timer[i].pauseFlag  = (BlzBitOr(Timer[i].pauseFlag, FLAG_TIMER_ALTERED))
+		return true
+	end
+	return false
+end
+
+--- Example: `doPeriodically(1/32, function(t) DestroyTimer(t) end)`
 ---@param period number
 ---@param func function
 function doPeriodicaly(period, func)
-	local t = CreateTimer()
-	TimerStart(t, period, true, function() func(t) DestroyTimer(GetExpiredTimer()) end)
+	TimerStart(CreateTimer(), period, true, function() func(GetExpiredTimer()) DestroyTimer(GetExpiredTimer()) end)
```

Comment:
I think last merge was incorrect.
Also, this `DestroyTimer()` shouldn't be here (my own stupidity), so please remove it.

Reply:
Will remove.